### DISCUSSION TITLE
The result of getSnapshot should be cached to avoid an infinite loop

### DISCUSSION THREAD

**Discussion by jcuenod** [pmndrs/zustand#1936] [Bug report] [answered]
I'm new to zustand. I've got my basic functionality working, but I'm seeing this warning coming from `react-dom` in the console:

> Warning: The result of getSnapshot should be cached to avoid an infinite loop

The stack trace is not very helpful, but I'm not directly using `getSnapshot` anywhere, so I assume this is a zustand bug. I guess that it has been introduced by https://github.com/facebook/react/issues/24508.

Is this a zustand bug or should I be scrutinizing my own code more?

**Comment by dai-shi** [comment]
Are you creating a fresh store in render function?

**Reply by jcuenod** [reply]
I have a `store.ts` file with:

```
export const useMyStore = create<MyState>((set) => ({
```

In a component file I import `useMyStore` and use it inside the render function:

```
const MyReactComponent = (props) => {
  const store = useMyStore((state) => ...
```

Is that not the correct pattern?

**Reply by dai-shi** [reply]
It looks good. Can you reproduce your issue with https://codesandbox.io/s/react-new ?

**Reply by jcuenod** [reply, marked as answer]
Thank you, I figured it out. The problem was that I was using `useX` callbacks to get around some eslint errors about the use of `useX` hooks outside react components. The actual solution was:

```
const myState = useMyStore.getState()
```

**Comment by shub1233** [comment]
my suggestion would be to use "useShallow" [Docs](https://zustand.docs.pmnd.rs/hooks/use-shallow#useshallow-%E2%9A%9B%EF%B8%8F)

const [blah] = useMyStore(useShallow((state) => [state.blah]));

**Reply by chaos-MK** [reply]
thank you ma man, that helped me a lot, even with using  AI I didn't figure out the problem

**Reply by ashuvssut** [reply]
The docs link is invalid. Here's the updated one: https://zustand.docs.pmnd.rs/reference/hooks/use-shallow

i did use `useShallow` as shown below, and i am still get the errors

```ts
import { useShallow } from 'zustand/react/shallow'
import { Keyboard } from 'react-native'

// inside the component ....

const gotoNextStep = useOnboarding(
	useShallow(
		(s) =>
			function gotoNextStepWithKeyboardDismiss() {
				Keyboard.dismiss()
				return s.gotoNextStep()
			}
	)
)
```

Here's the errors:
```
 ERROR  The result of getSnapshot should be cached to avoid an infinite loop
 ERROR  [Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.]
 ```
 
 ### Here's what works for me instead
 
 ```ts
 	const gotoNextStepFn = useOnboarding((s) => s.gotoNextStep)
 	const gotoNextStep = () => {
		Keyboard.dismiss()
		gotoNextStepFn()
	}
```

I use react 19.2. so that's why i didnt wrap `gotoNextStep` with `useCallback` -- and it works this way

I use zustand v5.0.12

I would have created the stackblitz for this, but i am on a tight deadline. i will create it when i get the time

**Comment by tylim88** [comment]
my case is 
`useStoreX(({ a, b }) => ({ a, b }))` triggers infinite rerender error
`useStoreX(state=>state)` and `useStoreX()` are fine
weird

**Reply by dai-shi** [reply]
try `useStoreX(useShallow(({ a, b }) => ({ a, b })))`

**Reply by tylim88** [reply]
useShallow works, thanks

**Reply by SOMONSOUM** [reply]
I still got `The result of getSnapshot should be cached to avoid an infinite loop`

My Store: 

```
import { RolesResponse } from "@/__generated__/graphql";
import { ErrorResponse } from "@/types";
import { StoreApi, UseBoundStore, create } from "zustand";

interface RolesState {
  roles: RolesResponse | null;
  error?: ErrorResponse | null;
  setRoles: (roles: RolesResponse | null) => void;
  setError: (error?: ErrorResponse | null) => void;
}

export const createRolesStore = (
  initialRoles: RolesResponse | null,
  initialError?: ErrorResponse | null
) =>
  create<RolesState>()((set) => ({
    roles: initialRoles,
    error: initialError,
    setRoles: (roles: RolesResponse | null) => set({ roles }),
    setError: (error?: ErrorResponse | null) => set({ error }),
  }));

export let useRolesStore: UseBoundStore<StoreApi<RolesState>>;

export const initializeRolesStore = (
  initialRoles: RolesResponse | null,
  initialError?: ErrorResponse | null
) => {
  useRolesStore = createRolesStore(initialRoles, initialError);
};
```

and this is what I use 
```
 const { roles, error } = useRolesStore(
    useShallow(({ roles, error }) => ({
      roles,
      error,
    }))
  );
  ```

**Reply by dai-shi** [reply]
If you change it to `useRolesStore((s) => s)`, does it avoid the error?

**Reply by SOMONSOUM** [reply]
> If you change it to `useRolesStore((s) => s)`, does it avoid the error?

No, it still get the same error unless I use Zustand 4.5.4; then it works fine.

This what I use with version 4.5.4

```
  const { roles, error } = useRolesStore((state) => ({
    roles: state.roles,
    error: state.error,
  }));
```

**Reply by dai-shi** [reply]
If `useRolesStore((s) => s)` still causes an error, `useShallow` doesn't help and it's a different issue. You may need to create a minimal reproduction with https://stackblitz.com.

**Reply by eric-d-johnson** [reply]
> try `useStoreX(useShallow(({ a, b }) => ({ a, b })))`

Is there a way to fix this at the store level, perhaps by creating the store like this? 
```
const useStoreX = createWithEqualityFn(
  (set) => ({
    a: ...,
    b: ...,
  }),
  shallow,
);

**Reply by eric-d-johnson** [reply]
Whoops, I see the reply below where you suggest this! 🤦‍♂️

**Reply by GauravJawalkar** [reply]
> > If you change it to `useRolesStore((s) => s)`, does it avoid the error?
> 
> No, it still get the same error unless I use Zustand 4.5.4; then it works fine.
> 
> This what I use with version 4.5.4
> 
> ```
>   const { roles, error } = useRolesStore((state) => ({
>     roles: state.roles,
>     error: state.error,
>   }));
> ```

Yepp worked in my case too

**Reply by DjonniStorm** [reply]
thanks, in my case this:
```
const { items, addItems } = useMyStore((state) => ({
        items: state.items,
        addItems: state.addItems,
}));
```    
caused an infinite loop

and this did not:
```
const { items, addItems } = useMyStore();
```

**Reply by tyx2000** [reply]
the same error here when you follow the zustand doc, and if just useStore() without state => ({...}) in it, the infinite loop is gone, but why?

**Reply by dbritto-dev** [reply]
@tyx2000 if you are returning a new object or a new array or a new ref that is causing an infinite loop so you must use `useShallow` or `shallow`. We recommend to use granular selector instead.

**Comment by sFrady20** [comment]
I'm having a similar issue after upgrading from 4.5.2 to 5.0.0

In my case I was doing:
```
  const xMove = app((x) => x.mouse.x / x.window.width - 0.5);
  const yMove = app((x) => x.mouse.y / x.window.height - 0.5);
 ```
I was able to confirm `app((x) => x.mouse.x);` and `app((x) => x.window.width);` each worked separately. but not when i try to use them both in the same selector.

I tried useShallow and it didn't work. (still getting infinite loop error)

I implemented a workaround by separating the accessors into different hooks:
```
  const mouseX = app((x) => x.mouse.x);
  const mouseY = app((x) => x.mouse.y);
  const winWidth = app((x) => x.window.width);
  const winHeight = app((x) => x.window.height);
  ...
  (then the rest of the math later)
```

I feel like pulling each variable individually is an unintended restriction though, and I'd personally prefer not to have to decide whether to use useShallow everytime FWIW

**Reply by dai-shi** [reply]
> I tried useShallow and it didn't work. (still getting infinite loop error)

That's sounds weird.

> I'd personally prefer not to have to decide whether to use useShallow everytime

For that, we introduced `createWithEqualityFn` which you can customize the default equalityFn and you can specify `shallow`. It's already in v4.

**Comment by Noitham** [comment]
Similar case with useShallow here:

I had this selector:

```
  const { selectedTheme, setSelectedTheme } = useStore(state => ({
    selectedTheme: state.selectedTheme,
    setSelectedTheme: state.setSelectedTheme,
  }));
```
Worked fine on v4, but on v5 I need to use `useShallow`.

**Reply by dai-shi** [reply]
Please use useShallow. v4 behavior seemed nice but it's not optimized.

**Reply by ShafSpecs** [reply]
Would `createWithEqualityFn` with `shallow` also do the trick?

**Reply by dbritto-dev** [reply]
@ShafSpecs yes

**Comment by dnguyenfs** [comment]
I have same problem when upgrade to v5, v4 working greate

**Reply by dai-shi** [reply]
v4 seems working, but it just hides this error.
Can you create a minimal reproduction with https://stackblitz.com/ ?

**Reply by Jumzeey** [reply]
i have the same issue

this is my store

```tsx
import { create } from 'zustand';

import {
  SellerInformation,
  VehicleInformation,
} from '@/types/Static/VehicleInformation';

// Define Zustand store type
type FormStore = {
  seller: SellerInformation;
  vehicle: VehicleInformation;
  setSeller: (seller: Partial<SellerInformation>) => void;
  setVehicle: (vehicle: Partial<VehicleInformation>) => void;
};

// Create Zustand store
export const useFormStore = create<FormStore>((set) => ({
  seller: {
    fullName: '',
    phone: '',
    email: '',
  },
  vehicle: {
    name: '',
    category: '',
    vin: '',
    id: '',
    address: '',
    registrationNumber: '',
    location: '',
  },
  setSeller: (seller) =>
    set((state) => ({ seller: { ...state.seller, ...seller } })),
  setVehicle: (vehicle) =>
    set((state) => ({ vehicle: { ...state.vehicle, ...vehicle } })),
}));
```

**Reply by dai-shi** [reply]
It looks good. Can you create a minimal reproduction with https://stackblitz.com/ ?

**Comment by Toshinaki** [comment]
same problem here.
I'm using the [slice pattern](https://zustand.docs.pmnd.rs/guides/typescript#slices-pattern)

whether adding `useShallow` or not, my `useStore` hook shows that `getSnapshot` error and causes infinite loops.

Downgrading to v4 solved the problem, without any code changes.

**Reply by dai-shi** [reply]
Can you create a minimal reproduction with https://stackblitz.com/ ?

**Reply by Nour-benmohamed-Git** [reply]
i have the same problem i did all possible solution without any result :(

**Reply by dbritto-dev** [reply]
@Nour-benmohamed-Git would you mind creating a minimal repro on stackblitz?

**Reply by Toshinaki** [reply]
@dai-shi @dbritto-dev 
do you guys have any template for us to start with?
I'd be happy to provide a reproduction but create everything from scratch is very frustrating.

**Reply by dai-shi** [reply]
We are working on a stater: #2913

Can use this for now: https://stackblitz.com/github/leweyse/zustand/tree/feat/starter-example/examples/starter

**Reply by Toshinaki** [reply]
@dai-shi 
here you go:
https://stackblitz.com/edit/github-omy9tt8q?file=src%2Fapp.tsx,package.json

without `useShallow`, `useAppStore` causes infinite loop.
but even with `useShallow`, this line also cause an infinite loop:
```ts
useShallow((state) => [
      state.test,
      state.test2,
      state.test2.items.map((item) => `-${item}`),  // <== this
])
```
removing this line and there's no infinite loop.

maybe I used this in a wrong way? v4 causes no error with the same code though.

---

also noticed the import of `useShallow`:
- `import { useShallow } from 'zustand/shallow'`
  https://zustand.docs.pmnd.rs/migrations/migrating-to-v5
- `import { useShallow } from 'zustand/react/shallow'`
   https://zustand.docs.pmnd.rs/guides/prevent-rerenders-with-use-shallow

**Reply by dai-shi** [reply]
`.map` creates a new array. 
```js
useShallow((state) => [
      state.test,
      [] // this doesn't work either
])
```
This works:
```js
useShallow((state) => 
      state.test2.items.map((item) => `-${item}`)
)
```

**Comment by wenerme** [comment]
if shallow is not enough, try `useDeepEqual`

```ts
export function useDeepEqual<S, U>(selector: (state: S) => U): (state: S) => U {
  // https://github.com/pmndrs/zustand/blob/main/src/react/shallow.ts
  const prev = useRef<U>(undefined);
  return (state) => {
    const next = selector(state);
    return deepEqual(prev.current, next) ? (prev.current as U) : (prev.current = next);
  };
}
```

**Comment by KestasVenslauskas** [comment]
In my case it was the new empty array generation inside selector:

Infifite loop code:
  ```
const state = useArticleStore(
    useShallow((state: ArticleState) => {
      const block = state.advancedCategories[id];
      if (block) {
        return {
          refreshing: block.isFetching && Boolean(block.items?.length),
          lastFetchTime: block.lastFetchTime,
//NOTE: new array generation
          items: block.items ?? [],
        };
      } else {
        return {
          refreshing: false,
          lastFetchTime: 0,
//NOTE: new array generation
          items: [],
        };
      }
    }),
  );
```


Fixed code:
```
  const emptyArray: HomeBlockType[] = [];
  const state = useArticleStore(
    useShallow((state: ArticleState) => {
      const block = state.advancedCategories[id];
      if (block) {
        return {
          refreshing: block.isFetching && Boolean(block.items?.length),
          lastFetchTime: block.lastFetchTime,
          items: block.items ?? emptyArray,
        };
      } else {
        return {
          refreshing: false,
          lastFetchTime: 0,
          items: emptyArray,
        };
      }
    }),
  );
```

**Reply by mayank1513** [reply]
Using `useShallow` is not a very good DX.

**Comment by mayank1513** [comment]
🚀 Hey folks! Give **Kosha** a try — [https://www.npmjs.com/package/kosha](https://www.npmjs.com/package/kosha)
It’s optimized out of the box — no need for `shallow` hooks or extra selectors.

👉 If you're using plain Zustand without middleware, migrating is super simple:
Just replace

```ts
import { create } from "zustand"
```

with

```ts
import { create } from "kosha"
```

✅ And if you’re using the TypeScript double-call syntax like `create()<T>(set => ({...}))`,
just switch to the simpler single-call version:

```ts
create<T>(set => ({...}))
```

That’s it — you’re good to go!

**Comment by mayank1513** [comment]
Thanks for the feedback. You are correct, it doesn't have plugin
eco-system. As I mentioned earlier, straightforward replacement is for most
users who don't use any middlewares.

Yes. Json.stringify works, if any specific concerns could you please let
me know. Thanks.



Mayank Kumar Chaudhari
Software Architect (Focus Front End) | Microsoft Certified Expert
https://mayank-chaudhari.vercel.app/

**Comment by mayank1513** [comment]
In most use cases, selectors return small objects or arrays.

Secondly, just double confirmed ` JSON.stringify({a: new Date()})` returns
` {"a":"2025-06-05T15:54:24.279Z"}`

Thanks for pointing out `Set` and `Map`. It's very rare. But still I shall
update this in documentation. Should not use `kosha` if you want to store
`Set` or `Map` to global store.


Mayank Kumar Chaudhari
Software Architect (Focus Front End) | Microsoft Certified Expert
https://mayank-chaudhari.vercel.app/

**Comment by mayank1513** [comment]
That will not happen as we are comparing results of the same selector
function - order is preserved.

Mayank Kumar Chaudhari
Software Architect (Focus Front End) | Microsoft Certified Expert
https://mayank-chaudhari.vercel.app/

**Comment by Mustafa-Alie** [comment]
the problem : 
If your state selector in react component return an object with with calculated values, this causes an infinite loop error 
This problem happens if you use Zustand v5

Root cause:
Since Zustand v5 adopted React’s **useSyncExternalStore** under the hood (to support concurrent rendering and React 18+ compatibility), it delegates all re-render logic to React — including snapshot comparison

solution:
1- Return primitive or existing state data (not an object or an array)
2- Use a constant (memoized) array
3- Use Zustand Vanilla Store or **createWithEqualityFn** with custom equality

my preferred solution:
assuming that you have an array of todos in your store.

function useTodosMeta() {
    const todos = useStore((state) => state.todos);
    return useMemo(
      () => todos.map((b) => ({ id: b.id, title: b.title })),
      [todos],
    );
const todosMeta = useTodosMeta();
console.log(todosMeta )

**Comment by Toshinaki** [comment]
Why not add the memorization (shallow & deepEquation) as a default?
since in a complicated app, in most cases what we need is a deeply nested object / array.
it feels verbose to add memorization & comparison manually every time.
**People use state management libraries for complex use cases.**

If all you want is just a simple primitive, there's really **no need to use a 3rd party state management**.
context and/or useState is more than enough for simple use cases.
